Bottom: 885a886433bd704c8a35630ef2947a798a36e322
Top:    8b04fee0da3691a944fa466d2383c1d650beb5fd
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2014-06-06 12:15:14 +0100

Refresh of continue-tinkering-with-3d

---

diff --git a/src/hid/gtk/gui-dialog.c b/src/hid/gtk/gui-dialog.c
index 56e0d25..9344abd 100644
--- a/src/hid/gtk/gui-dialog.c
+++ b/src/hid/gtk/gui-dialog.c
@@ -231,6 +231,7 @@ ghid_dialog_close_confirm ()
                           GTK_STOCK_CANCEL,          GTK_RESPONSE_CANCEL,
                           GTK_STOCK_SAVE,            GTK_RESPONSE_YES,
                           NULL);
+  gtk_dialog_set_default_response (GTK_DIALOG (dialog), GTK_RESPONSE_YES);
 
   /* Set the alternative button order (ok, cancel, help) for other systems */
   gtk_dialog_set_alternative_button_order(GTK_DIALOG(dialog),
diff --git a/src/hid/gtk/object3d.c b/src/hid/gtk/object3d.c
index 1e2c7e3..13bab94 100644
--- a/src/hid/gtk/object3d.c
+++ b/src/hid/gtk/object3d.c
@@ -35,7 +35,7 @@
 
 #define HACK_BOARD_THICKNESS MM_TO_COORD(1.6)
 
-static object3d *object3d_test_object = NULL;
+static GList *object3d_test_objects = NULL;
 
 static void
 print_edge_id (edge_ref e)
@@ -61,8 +61,8 @@ debug_print_edge (edge_ref e, void *data)
 void
 object3d_test_init (void)
 {
-  //object3d_test_object = object3d_create_test_cube ();
-  object3d_test_object = object3d_from_board_outline ();
+  //object3d_test_objects = g_list_append (NULL, object3d_create_test_cube ());
+  object3d_test_objects = object3d_from_board_outline ();
   object3d_test_board_outline ();
 }
 
@@ -240,13 +240,19 @@ draw_quad_edge (edge_ref e, void *data)
   glEnd ();
 }
 
+static void
+object3d_draw_debug_single (object3d *object, void *user_data)
+{
+  g_return_if_fail (object->edges != NULL);
+
+//  quad_enum ((edge_ref)object->edges->data, draw_quad_edge, NULL);
+  g_list_foreach (object->edges, (GFunc)draw_quad_edge, NULL);
+}
+
 void
 object3d_draw_debug (void)
 {
-  g_return_if_fail (object3d_test_object->edges != NULL);
-
-//  quad_enum ((edge_ref)object3d_test_object->edges->data, draw_quad_edge, NULL);
-  g_list_foreach (object3d_test_object->edges, (GFunc)draw_quad_edge, NULL);
+  g_list_foreach (object3d_test_objects, (GFunc)object3d_draw_debug_single, NULL);
 }
 
 /*********************************************************************************************************/
@@ -298,6 +304,7 @@ object3d_export_to_step (object3d *object, char *filename)
   struct tm utc;
   int next_step_identifier;
   int geometric_representation_context_identifier;
+  int shape_representation_identifier;
   int brep_identifier;
   int pcb_shell_identifier;
   int brep_style_identifier;
@@ -368,12 +375,13 @@ object3d_export_to_step (object3d *object, char *filename)
               "#16 =( NAMED_UNIT ( * ) PLANE_ANGLE_UNIT ( ) SI_UNIT ( $, .RADIAN. ) );\n"
               "#17 =( NAMED_UNIT ( * ) SI_UNIT ( $, .STERADIAN. ) SOLID_ANGLE_UNIT ( ) );\n"
               "#18 =( GEOMETRIC_REPRESENTATION_CONTEXT ( 3 ) GLOBAL_UNCERTAINTY_ASSIGNED_CONTEXT ( ( #14 ) ) GLOBAL_UNIT_ASSIGNED_CONTEXT ( ( #15, #16, #17 ) ) REPRESENTATION_CONTEXT ( 'NONE', 'WORKASPACE' ) );\n");
+  geometric_representation_context_identifier = 18;
 
-  fprintf (f, "#19 = ADVANCED_BREP_SHAPE_REPRESENTATION ( '%s', ( /* Manifold_solid_brep */ #21, #13 ), #18 ) ;\n"
-              "#20 = SHAPE_DEFINITION_REPRESENTATION ( #9, #19 ) ;\n",
-              "test_pcb_absr_name");
+  /* Save a place for the advanced_brep_shape_representation identifier */
+  next_step_identifier = 19;
+  shape_representation_identifier = next_step_identifier++;
 
-  geometric_representation_context_identifier = 18;
+  fprintf (f, "#20 = SHAPE_DEFINITION_REPRESENTATION ( #9, #%i ) ;\n", shape_representation_identifier);
 
   /* Save a place for the brep identifier */
   next_step_identifier = 21;
@@ -557,7 +565,6 @@ object3d_export_to_step (object3d *object, char *filename)
         contour3d *contour = contour_iter->data;
         edge_ref edge;
 
-        /* XXX: FWD / BWD NEEDS TO BE FUDGED IN HERE PERHAPS? */
         fprintf (f, "#%i = EDGE_LOOP ( 'NONE', ", next_step_identifier);
 
         /* Emit the edges.. */
@@ -632,10 +639,12 @@ object3d_export_to_step (object3d *object, char *filename)
   fprintf (f, ", #%i ) ;\n", geometric_representation_context_identifier);
   next_step_identifier = next_step_identifier + 1;
 
+  fprintf (f, "#%i = ADVANCED_BREP_SHAPE_REPRESENTATION ( '%s', ( #%i, #13 ), #%i ) ;\n",
+           shape_representation_identifier, "test_pcb_absr_name", brep_identifier, geometric_representation_context_identifier);
 
+#undef ORIENTED_EDGE_IDENTIFIER
 #undef FWD
 #undef REV
-#undef ORIENTED_EDGE_IDENTIFIER
 
   fprintf (f, "ENDSEC;\n" );
   fprintf (f, "END-ISO-10303-21;\n" );
@@ -643,13 +652,15 @@ object3d_export_to_step (object3d *object, char *filename)
   fclose (f);
 }
 
-object3d *
+GList *
 object3d_from_board_outline (void)
 {
+  GList *board_objects = NULL;
   object3d *board_object;
   appearance *board_appearance;
   appearance *top_bot_appearance;
-  POLYAREA *outline;
+  POLYAREA *board_outline;
+  POLYAREA *pa;
   PLINE *contour;
   PLINE *ct;
   int ncontours;
@@ -662,273 +673,279 @@ object3d_from_board_outline (void)
   int offset_in_ct;
   int ct_npoints;
 
-  outline = board_outline_poly (true);
-  //outline = board_outline_poly (false); /* (FOR NOW - just the outline, no holes) */
-  ncontours = 0;
-  npoints = 0;
+  board_outline = board_outline_poly (true);
 
-  /* XXX: There can be more than one contour, but for now we restrict ourselves to the first one */
-  contour = outline->contours;
+  /* Loop over all board outline pieces */
+  pa = board_outline;
+  do {
 
-  ct = contour;
-  while (ct != NULL) {
-    ncontours ++;
-    npoints += get_contour_npoints (ct);
-    ct = ct->next;
-  }
+    contour = pa->contours;
+    ncontours = 0;
+    npoints = 0;
 
-  /* We know how many edges and vertices we need now...
-   *
-   * let n = npoints
-   * bodies = 1             (FOR NOW - just the first board outline)
-   * vertices = 2n          (n-top, n-bottom)
-   * edges = 3n             (n-top, n-bottom, n-sides)
-   * faces = 2 + n          (1-top, 1-bottom, n-sides)
-   *
-   * holes = 0              (FOR NOW - just the outline, no holes)
-   * holes = ncontours - 1  (LATER)
-   */
-
-  board_object = make_object3d (PCB->Name);
-  board_appearance = make_appearance ();
-  top_bot_appearance = make_appearance ();
-  appearance_set_color (board_appearance,   1.0, 1.0, 0.6);
-  appearance_set_color (top_bot_appearance, 0.2, 0.8, 0.2);
-
-  object3d_set_appearance (board_object, board_appearance);
-
-  vertices = malloc (sizeof (vertex3d *) * 2 * npoints);
-  edges    = malloc (sizeof (edge_ref  ) * 3 * npoints);
-  faces    = malloc (sizeof (face3d *) * (2 + npoints));
+    ct = contour;
+    while (ct != NULL) {
+      ncontours ++;
+      npoints += get_contour_npoints (ct);
+      ct = ct->next;
+    }
 
-  /* Define the vertices */
-  ct = contour;
-  start_of_ct = 0;
-  offset_in_ct = 0;
-  ct_npoints = get_contour_npoints (ct);
+    /* We know how many edges and vertices we need now...
+     *
+     * let n = npoints
+     * bodies = 1             (FOR NOW - just the first board outline)
+     * vertices = 2n          (n-top, n-bottom)
+     * edges = 3n             (n-top, n-bottom, n-sides)
+     * faces = 2 + n          (1-top, 1-bottom, n-sides)
+     *
+     * holes = 0              (FOR NOW - just the outline, no holes)
+     * holes = ncontours - 1  (LATER)
+     */
 
-  for (i = 0; i < npoints; i++, offset_in_ct++) {
-    double x1, y1;
+    board_object = make_object3d (PCB->Name);
+    board_appearance = make_appearance ();
+    top_bot_appearance = make_appearance ();
+    appearance_set_color (board_appearance,   1.0, 1.0, 0.6);
+    appearance_set_color (top_bot_appearance, 0.2, 0.8, 0.2);
 
-    /* Update which contour we're looking at */
-    if (offset_in_ct == ct_npoints) {
-      offset_in_ct = 0;
-      ct = ct->next;
-      ct_npoints = get_contour_npoints (ct);
-    }
+    object3d_set_appearance (board_object, board_appearance);
 
-    get_contour_coord_n_in_mm (ct, offset_in_ct, &x1, &y1);
-    vertices[i]           = make_vertex3d (x1, y1, -COORD_TO_MM (HACK_BOARD_THICKNESS)); /* Bottom */
-    vertices[npoints + i] = make_vertex3d (x1, y1, 0);                                   /* Top */
+    vertices = malloc (sizeof (vertex3d *) * 2 * npoints);
+    edges    = malloc (sizeof (edge_ref  ) * 3 * npoints);
+    faces    = malloc (sizeof (face3d *) * (2 + npoints));
 
-    object3d_add_vertex (board_object, vertices[i]);
-    object3d_add_vertex (board_object, vertices[npoints + i]);
-  }
+    /* Define the vertices */
+    ct = contour;
+    start_of_ct = 0;
+    offset_in_ct = 0;
+    ct_npoints = get_contour_npoints (ct);
 
-  /* Define the edges */
-  for (i = 0; i < 3 * npoints; i++) {
-    edges[i] = make_edge ();
-    UNDIR_DATA (edges[i]) = make_edge_info ();
-    object3d_add_edge (board_object, edges[i]);
-  }
+    for (i = 0; i < npoints; i++, offset_in_ct++) {
+      double x1, y1;
 
-  /* Define the faces */
-  for (i = 0; i < npoints; i++) {
-    faces[i] = make_face3d ();
+      /* Update which contour we're looking at */
+      if (offset_in_ct == ct_npoints) {
+        offset_in_ct = 0;
+        ct = ct->next;
+        ct_npoints = get_contour_npoints (ct);
+      }
 
-    object3d_add_face (board_object, faces[i]);
-    /* Pick one of the upright edges which is within this face outer contour loop, and link it to the face */
-    face3d_add_contour (faces[i], make_contour3d (SYM(edges[2 * npoints + i])));
-  }
+      get_contour_coord_n_in_mm (ct, offset_in_ct, &x1, &y1);
+      vertices[i]           = make_vertex3d (x1, y1, -COORD_TO_MM (HACK_BOARD_THICKNESS)); /* Bottom */
+      vertices[npoints + i] = make_vertex3d (x1, y1, 0);                                   /* Top */
 
-  faces[npoints] = make_face3d (); /* bottom_face */
-  face3d_set_normal (faces[npoints], 0., 0., -1.);
-  face3d_set_appearance (faces[npoints], top_bot_appearance);
-  object3d_add_face (board_object, faces[npoints]);
-
-  faces[npoints + 1] = make_face3d (); /* top_face */
-  face3d_set_normal (faces[npoints + 1], 0., 0., 1.);
-  face3d_set_appearance (faces[npoints + 1], top_bot_appearance);
-  object3d_add_face (board_object, faces[npoints + 1]);
-
-  /* Pick the first bottom / top edge within the bottom / top face outer contour loop, and link it to the face */
-  face3d_add_contour (faces[npoints], make_contour3d (SYM(edges[0])));
-  face3d_add_contour (faces[npoints + 1], make_contour3d (edges[npoints]));
-
-  ct = contour;
-  start_of_ct = 0;
-  offset_in_ct = 0;
-  ct_npoints = get_contour_npoints (ct);
-
-  for (i = 0; i < npoints; i++, offset_in_ct++) {
-    int next_i_around_ct;
-    int prev_i_around_ct;
-
-    /* Update which contour we're looking at */
-    if (offset_in_ct == ct_npoints) {
-      start_of_ct = i;
-      offset_in_ct = 0;
-      ct = ct->next;
-      ct_npoints = get_contour_npoints (ct);
+      object3d_add_vertex (board_object, vertices[i]);
+      object3d_add_vertex (board_object, vertices[npoints + i]);
+    }
 
-      /* If there is more than one contour, it will be an inner contour of the bottom and top faces. Refer to it here */
-      face3d_add_contour (faces[npoints], make_contour3d (SYM(edges[i])));
-      face3d_add_contour (faces[npoints + 1], make_contour3d (edges[npoints + i]));
+    /* Define the edges */
+    for (i = 0; i < 3 * npoints; i++) {
+      edges[i] = make_edge ();
+      UNDIR_DATA (edges[i]) = make_edge_info ();
+      object3d_add_edge (board_object, edges[i]);
     }
 
-    next_i_around_ct = start_of_ct + (offset_in_ct + 1) % ct_npoints;
-    prev_i_around_ct = start_of_ct + (offset_in_ct + ct_npoints - 1) % ct_npoints;
-
-    /* Setup the face normals for the edges along the contour extrusion (top and bottom are handled separaetely) */
-    /* Define the (non-normalized) face normal to point to the outside of the contour */
-    face3d_set_normal (faces[i], (vertices[next_i_around_ct]->y - vertices[i]->y),
-                                -(vertices[next_i_around_ct]->x - vertices[i]->x), 0.);
-
-    /* Assign the appropriate vertex geometric data to each edge end */
-    ODATA (edges[              i]) = vertices[0 * npoints + i];
-    DDATA (edges[              i]) = vertices[0 * npoints + next_i_around_ct];
-    ODATA (edges[1 * npoints + i]) = vertices[1 * npoints + i];
-    DDATA (edges[1 * npoints + i]) = vertices[1 * npoints + next_i_around_ct];
-    ODATA (edges[2 * npoints + i]) = vertices[0 * npoints + i];
-    DDATA (edges[2 * npoints + i]) = vertices[1 * npoints + i];
-    LDATA (edges[              i]) = faces[i];
-    RDATA (edges[              i]) = faces[npoints];
-    LDATA (edges[1 * npoints + i]) = faces[npoints + 1];
-    RDATA (edges[1 * npoints + i]) = faces[i];
-    LDATA (edges[2 * npoints + i]) = faces[prev_i_around_ct];
-    RDATA (edges[2 * npoints + i]) = faces[i];
-
-    /* NB: Contours are counter clockwise in XY plane.
-     *     edges[          0-npoints-1] are the base of the extrusion, following in the counter clockwise order
-     *     edges[1*npoints-2*npoints-1] are the top  of the extrusion, following in the counter clockwise order
-     *     edges[2*npoints-3*npoints-1] are the upright edges, oriented from bottom to top
-     */
+    /* Define the faces */
+    for (i = 0; i < npoints; i++) {
+      faces[i] = make_face3d ();
+
+      object3d_add_face (board_object, faces[i]);
+      /* Pick one of the upright edges which is within this face outer contour loop, and link it to the face */
+      face3d_add_contour (faces[i], make_contour3d (SYM(edges[2 * npoints + i])));
+    }
+
+    faces[npoints] = make_face3d (); /* bottom_face */
+    face3d_set_normal (faces[npoints], 0., 0., -1.);
+    face3d_set_appearance (faces[npoints], top_bot_appearance);
+    object3d_add_face (board_object, faces[npoints]);
+
+    faces[npoints + 1] = make_face3d (); /* top_face */
+    face3d_set_normal (faces[npoints + 1], 0., 0., 1.);
+    face3d_set_appearance (faces[npoints + 1], top_bot_appearance);
+    object3d_add_face (board_object, faces[npoints + 1]);
+
+    /* Pick the first bottom / top edge within the bottom / top face outer contour loop, and link it to the face */
+    face3d_add_contour (faces[npoints], make_contour3d (SYM(edges[0])));
+    face3d_add_contour (faces[npoints + 1], make_contour3d (edges[npoints]));
+
+    ct = contour;
+    start_of_ct = 0;
+    offset_in_ct = 0;
+    ct_npoints = get_contour_npoints (ct);
+
+    for (i = 0; i < npoints; i++, offset_in_ct++) {
+      int next_i_around_ct;
+      int prev_i_around_ct;
+
+      /* Update which contour we're looking at */
+      if (offset_in_ct == ct_npoints) {
+        start_of_ct = i;
+        offset_in_ct = 0;
+        ct = ct->next;
+        ct_npoints = get_contour_npoints (ct);
+
+        /* If there is more than one contour, it will be an inner contour of the bottom and top faces. Refer to it here */
+        face3d_add_contour (faces[npoints], make_contour3d (SYM(edges[i])));
+        face3d_add_contour (faces[npoints + 1], make_contour3d (edges[npoints + i]));
+      }
+
+      next_i_around_ct = start_of_ct + (offset_in_ct + 1) % ct_npoints;
+      prev_i_around_ct = start_of_ct + (offset_in_ct + ct_npoints - 1) % ct_npoints;
+
+      /* Setup the face normals for the edges along the contour extrusion (top and bottom are handled separaetely) */
+      /* Define the (non-normalized) face normal to point to the outside of the contour */
+      face3d_set_normal (faces[i], (vertices[next_i_around_ct]->y - vertices[i]->y),
+                                  -(vertices[next_i_around_ct]->x - vertices[i]->x), 0.);
+
+      /* Assign the appropriate vertex geometric data to each edge end */
+      ODATA (edges[              i]) = vertices[0 * npoints + i];
+      DDATA (edges[              i]) = vertices[0 * npoints + next_i_around_ct];
+      ODATA (edges[1 * npoints + i]) = vertices[1 * npoints + i];
+      DDATA (edges[1 * npoints + i]) = vertices[1 * npoints + next_i_around_ct];
+      ODATA (edges[2 * npoints + i]) = vertices[0 * npoints + i];
+      DDATA (edges[2 * npoints + i]) = vertices[1 * npoints + i];
+      LDATA (edges[              i]) = faces[i];
+      RDATA (edges[              i]) = faces[npoints];
+      LDATA (edges[1 * npoints + i]) = faces[npoints + 1];
+      RDATA (edges[1 * npoints + i]) = faces[i];
+      LDATA (edges[2 * npoints + i]) = faces[prev_i_around_ct];
+      RDATA (edges[2 * npoints + i]) = faces[i];
+
+      /* NB: Contours are counter clockwise in XY plane.
+       *     edges[          0-npoints-1] are the base of the extrusion, following in the counter clockwise order
+       *     edges[1*npoints-2*npoints-1] are the top  of the extrusion, following in the counter clockwise order
+       *     edges[2*npoints-3*npoints-1] are the upright edges, oriented from bottom to top
+       */
+
+      /* Link edges orbiting around each bottom vertex i (0 <= i < npoints) */
+      splice (edges[i], edges[2 * npoints + i]);
+      splice (edges[2 * npoints + i], SYM(edges[prev_i_around_ct]));
+
+      /* Link edges orbiting around each top vertex (npoints + i) (0 <= i < npoints) */
+      splice (SYM(edges[2 * npoints + i]), edges[npoints + i]);
+      splice (edges[npoints + i], SYM(edges[npoints + prev_i_around_ct]));
 
-    /* Link edges orbiting around each bottom vertex i (0 <= i < npoints) */
-    splice (edges[i], edges[2 * npoints + i]);
-    splice (edges[2 * npoints + i], SYM(edges[prev_i_around_ct]));
+      if (ct->is_round) {
 
-    /* Link edges orbiting around each top vertex (npoints + i) (0 <= i < npoints) */
-    splice (SYM(edges[2 * npoints + i]), edges[npoints + i]);
-    splice (edges[npoints + i], SYM(edges[npoints + prev_i_around_ct]));
+        face3d_set_cylindrical (faces[i], COORD_TO_MM (ct->cx), COORD_TO_MM (ct->cy), 0., /* A point on the axis of the cylinder */
+                                          0., 0., 1.,                                     /* Direction of the cylindrical axis */
+                                          COORD_TO_MM (ct->radius));
+        face3d_set_normal (faces[i], 1., 0., 0.);  /* A normal to the axis direction */
+                                  /* XXX: ^^^ Could line this up with the direction to the vertex in the corresponding circle edge */
 
-    if (ct->is_round) {
 
-      face3d_set_cylindrical (faces[i], COORD_TO_MM (ct->cx), COORD_TO_MM (ct->cy), 0., /* A point on the axis of the cylinder */
-                                        0., 0., 1.,                                     /* Direction of the cylindrical axis */
-                                        COORD_TO_MM (ct->radius));
-      face3d_set_normal (faces[i], 1., 0., 0.);  /* A normal to the axis direction */
-                                /* XXX: ^^^ Could line this up with the direction to the vertex in the corresponding circle edge */
+        edge_info_set_round (UNDIR_DATA (edges[i]),
+                             COORD_TO_MM (ct->cx), COORD_TO_MM (ct->cy), -COORD_TO_MM (HACK_BOARD_THICKNESS), /* Center of circle */
+                             0., 0., -1., /* Normal */ COORD_TO_MM (ct->radius)); /* NORMAL POINTING TO -VE Z MAKES CIRCLE CLOCKWISE */
+        edge_info_set_round (UNDIR_DATA (edges[npoints + i]),
+                             COORD_TO_MM (ct->cx), COORD_TO_MM (ct->cy), 0., /* Center of circle */
+                             0., 0., -1., /* Normal */ COORD_TO_MM (ct->radius)); /* NORMAL POINTING TO -VE Z MAKES CIRCLE CLOCKWISE */
+        edge_info_set_stitch (UNDIR_DATA (edges[2 * npoints + i]));
+      }
 
+    }
 
-      edge_info_set_round (UNDIR_DATA (edges[i]),
-                           COORD_TO_MM (ct->cx), COORD_TO_MM (ct->cy), -COORD_TO_MM (HACK_BOARD_THICKNESS), /* Center of circle */
-                           0., 0., -1., /* Normal */ COORD_TO_MM (ct->radius)); /* NORMAL POINTING TO -VE Z MAKES CIRCLE CLOCKWISE */
-      edge_info_set_round (UNDIR_DATA (edges[npoints + i]),
-                           COORD_TO_MM (ct->cx), COORD_TO_MM (ct->cy), 0., /* Center of circle */
-                           0., 0., -1., /* Normal */ COORD_TO_MM (ct->radius)); /* NORMAL POINTING TO -VE Z MAKES CIRCLE CLOCKWISE */
-      edge_info_set_stitch (UNDIR_DATA (edges[2 * npoints + i]));
+    if (1) {
+      /* Cylinder centers on 45x45mm, stitch vertex is at 40x45mm. Radius is thus 5mm */
+
+      edge_ref cylinder_edges[3];
+      vertex3d *cylinder_vertices[2];
+      face3d *cylinder_faces[2];
+
+      /* Edge on top of board */
+      cylinder_edges[0] = make_edge ();
+      UNDIR_DATA (cylinder_edges[0]) = make_edge_info ();
+      edge_info_set_round (UNDIR_DATA (cylinder_edges[0]),
+                           45., 45., 0., /* Center of circle */
+                            0.,  0., 1., /* Normal */
+                            5.);         /* Radius */
+      object3d_add_edge (board_object, cylinder_edges[0]);
+
+      /* Edge on top of cylinder */
+      cylinder_edges[1] = make_edge ();
+      UNDIR_DATA (cylinder_edges[1]) = make_edge_info ();
+      edge_info_set_round (UNDIR_DATA (cylinder_edges[1]),
+                           45., 45., 10., /* Center of circle */
+                            0.,  0., 1.,  /* Normal */
+                            5.);          /* Radius */
+      object3d_add_edge (board_object, cylinder_edges[1]);
+
+      /* Edge stitching cylinder */
+      cylinder_edges[2] = make_edge ();
+      UNDIR_DATA (cylinder_edges[2]) = make_edge_info ();
+      edge_info_set_stitch (UNDIR_DATA (cylinder_edges[2]));
+      object3d_add_edge (board_object, cylinder_edges[2]);
+
+      /* Vertex on board top surface */
+      cylinder_vertices[0] = make_vertex3d (40., 45., 0.); /* Bottom */
+      object3d_add_vertex (board_object, cylinder_vertices[0]);
+
+      /* Vertex on cylinder top surface */
+      cylinder_vertices[1] = make_vertex3d (40., 45., 10.); /* Top */
+      object3d_add_vertex (board_object, cylinder_vertices[1]);
+
+      /* Cylindrical face */
+      cylinder_faces[0] = make_face3d ();
+      face3d_set_cylindrical (cylinder_faces[0], 45., 45., 0., /* A point on the axis of the cylinder */
+                                        0., 0., 1.,            /* Direction of the cylindrical axis */
+                                        5.);                   /* Radius of cylinder */
+      face3d_set_flip_orientation (cylinder_faces[0]); /* XXX: HACK TO IDENTIFY THIS AS AN OUTSIDE CYLINDRICAL SURFACE, NOT INTERNAL AS ASSUMED OTHERWISE */
+      face3d_set_normal (cylinder_faces[0], 1., 0., 0.);       /* A normal to the axis direction */
+                                   /* XXX: ^^^ Could line this up with the direction to the vertex in the corresponding circle edge */
+      object3d_add_face (board_object, cylinder_faces[0]);
+      face3d_add_contour (cylinder_faces[0], make_contour3d (cylinder_edges[0]));
+
+      /* Top face of cylinder */
+      cylinder_faces[1] = make_face3d (); /* top face of cylinder */
+      face3d_set_normal (cylinder_faces[1], 0., 0., 1.);
+      face3d_set_appearance (cylinder_faces[1], top_bot_appearance);
+      object3d_add_face (board_object, cylinder_faces[1]);
+      face3d_add_contour (cylinder_faces[1], make_contour3d (cylinder_edges[1]));
+
+      /* Splice onto board */
+      face3d_add_contour (faces[npoints + 1], make_contour3d (SYM(cylinder_edges[0])));
+
+      /* Assign the appropriate vertex geometric data to each edge end */
+      ODATA (cylinder_edges[0]) = cylinder_vertices[0];
+      DDATA (cylinder_edges[0]) = cylinder_vertices[0];
+      ODATA (cylinder_edges[1]) = cylinder_vertices[1];
+      DDATA (cylinder_edges[1]) = cylinder_vertices[1];
+      ODATA (cylinder_edges[2]) = cylinder_vertices[0];
+      DDATA (cylinder_edges[2]) = cylinder_vertices[1];
+      LDATA (cylinder_edges[0]) = cylinder_faces[0];
+      RDATA (cylinder_edges[0]) = faces[npoints + 1]; /* TOP OF BOARD FACE */
+      LDATA (cylinder_edges[1]) = cylinder_faces[1];
+      RDATA (cylinder_edges[1]) = cylinder_faces[0];
+      LDATA (cylinder_edges[2]) = cylinder_faces[0];
+      RDATA (cylinder_edges[2]) = cylinder_faces[0];
+
+      /* Splice things together.... */
+
+      /* Link edges orbiting the cylinder bottom vertex */
+      splice (cylinder_edges[0], cylinder_edges[2]);
+      splice (cylinder_edges[2], SYM(cylinder_edges[0]));
+
+      /* Link edges orbiting the cylinder top vertex */
+      splice (SYM(cylinder_edges[2]), cylinder_edges[1]);
+      splice (cylinder_edges[1], SYM(cylinder_edges[1]));
     }
 
-  }
+    board_objects = g_list_append (board_objects, board_object);
 
-  if (1) {
-    /* Cylinder centers on 45x45mm, stitch vertex is at 40x45mm. Radius is thus 5mm */
-
-    edge_ref cylinder_edges[3];
-    vertex3d *cylinder_vertices[2];
-    face3d *cylinder_faces[2];
-
-    /* Edge on top of board */
-    cylinder_edges[0] = make_edge ();
-    UNDIR_DATA (cylinder_edges[0]) = make_edge_info ();
-    edge_info_set_round (UNDIR_DATA (cylinder_edges[0]),
-                         45., 45., 0., /* Center of circle */
-                          0.,  0., 1., /* Normal */
-                          5.);         /* Radius */
-    object3d_add_edge (board_object, cylinder_edges[0]);
-
-    /* Edge on top of cylinder */
-    cylinder_edges[1] = make_edge ();
-    UNDIR_DATA (cylinder_edges[1]) = make_edge_info ();
-    edge_info_set_round (UNDIR_DATA (cylinder_edges[1]),
-                         45., 45., 10., /* Center of circle */
-                          0.,  0., 1.,  /* Normal */
-                          5.);          /* Radius */
-    object3d_add_edge (board_object, cylinder_edges[1]);
-
-    /* Edge stitching cylinder */
-    cylinder_edges[2] = make_edge ();
-    UNDIR_DATA (cylinder_edges[2]) = make_edge_info ();
-    edge_info_set_stitch (UNDIR_DATA (cylinder_edges[2]));
-    object3d_add_edge (board_object, cylinder_edges[2]);
-
-    /* Vertex on board top surface */
-    cylinder_vertices[0] = make_vertex3d (40., 45., 0.); /* Bottom */
-    object3d_add_vertex (board_object, cylinder_vertices[0]);
-
-    /* Vertex on cylinder top surface */
-    cylinder_vertices[1] = make_vertex3d (40., 45., 10.); /* Top */
-    object3d_add_vertex (board_object, cylinder_vertices[1]);
-
-    /* Cylindrical face */
-    cylinder_faces[0] = make_face3d ();
-    face3d_set_cylindrical (cylinder_faces[0], 45., 45., 0., /* A point on the axis of the cylinder */
-                                      0., 0., 1.,            /* Direction of the cylindrical axis */
-                                      5.);                   /* Radius of cylinder */
-    face3d_set_flip_orientation (cylinder_faces[0]); /* XXX: HACK TO IDENTIFY THIS AS AN OUTSIDE CYLINDRICAL SURFACE, NOT INTERNAL AS ASSUMED OTHERWISE */
-    face3d_set_normal (cylinder_faces[0], 1., 0., 0.);       /* A normal to the axis direction */
-                                 /* XXX: ^^^ Could line this up with the direction to the vertex in the corresponding circle edge */
-    face3d_set_appearance (cylinder_faces[0], top_bot_appearance);
-    object3d_add_face (board_object, cylinder_faces[0]);
-    face3d_add_contour (cylinder_faces[0], make_contour3d (cylinder_edges[0]));
-
-    /* Top face of cylinder */
-    cylinder_faces[1] = make_face3d (); /* top face of cylinder */
-    face3d_set_normal (cylinder_faces[1], 0., 0., 1.);
-    face3d_set_appearance (cylinder_faces[1], top_bot_appearance);
-    object3d_add_face (board_object, cylinder_faces[1]);
-    face3d_add_contour (cylinder_faces[1], make_contour3d (cylinder_edges[1]));
-
-    /* Splice onto board */
-    face3d_add_contour (faces[npoints + 1], make_contour3d (SYM(cylinder_edges[0])));
-
-    /* Assign the appropriate vertex geometric data to each edge end */
-    ODATA (cylinder_edges[0]) = cylinder_vertices[0];
-    DDATA (cylinder_edges[0]) = cylinder_vertices[0];
-    ODATA (cylinder_edges[1]) = cylinder_vertices[1];
-    DDATA (cylinder_edges[1]) = cylinder_vertices[1];
-    ODATA (cylinder_edges[2]) = cylinder_vertices[0];
-    DDATA (cylinder_edges[2]) = cylinder_vertices[1];
-    LDATA (cylinder_edges[0]) = cylinder_faces[0];
-    RDATA (cylinder_edges[0]) = faces[npoints + 1]; /* TOP OF BOARD FACE */
-    LDATA (cylinder_edges[1]) = cylinder_faces[1];
-    RDATA (cylinder_edges[1]) = cylinder_faces[0];
-    LDATA (cylinder_edges[2]) = cylinder_faces[0];
-    RDATA (cylinder_edges[2]) = cylinder_faces[0];
-
-    /* Splice things together.... */
-
-    /* Link edges orbiting the cylinder bottom vertex */
-    splice (cylinder_edges[0], cylinder_edges[2]);
-    splice (cylinder_edges[2], SYM(cylinder_edges[0]));
-
-    /* Link edges orbiting the cylinder top vertex */
-    splice (SYM(cylinder_edges[2]), cylinder_edges[1]);
-    splice (cylinder_edges[1], SYM(cylinder_edges[1]));
-  }
+  } while (pa = pa->f, pa != board_outline);
 
-  poly_Free (&outline);
+  poly_Free (&board_outline);
 
-  return board_object;
+  return board_objects;
 }
 
 void
 object3d_test_board_outline (void)
 {
-  object3d *board_outline;
+  GList *board_outlines;
+
+  board_outlines = object3d_from_board_outline ();
+  object3d_export_to_step (board_outlines->data, "object3d_test.step");
 
-  board_outline = object3d_from_board_outline ();
-  object3d_export_to_step (board_outline, "object3d_test.step");
-  destroy_object3d (board_outline);
+  g_list_free_full (board_outlines, (GDestroyNotify)destroy_object3d);
 }
diff --git a/src/hid/gtk/object3d.h b/src/hid/gtk/object3d.h
index 8d081a2..ecdc721 100644
--- a/src/hid/gtk/object3d.h
+++ b/src/hid/gtk/object3d.h
@@ -16,6 +16,6 @@ void object3d_add_edge (object3d *object, edge_ref edge);
 void object3d_add_vertex (object3d *object, vertex3d *vertex);
 void object3d_add_face (object3d *object, face3d *face);
 object3d *object3d_create_test_cube (void);
-object3d *object3d_from_board_outline (void);
+GList *object3d_from_board_outline (void);
 void object3d_export_to_step (object3d *object, char *filename);
 void object3d_test_board_outline (void);
